feat(cache) Add a django cache adapter that reconnects

markstory opened the pull request:
We often see load-imbalances in our memcache clusters. Application containers connect through a pool of twemproxy instances, and because application containers are long-lived and only connect on startup, we can end up with imbalanced load.

By periodically disconnecting and reconnecting to twemproxy we have a better chance of getting more even load distribution. It doesn't look like we have memcache available in CI right now, and writing tests with only mocks would yield tautological tests.

Refs PRODENG-702

markstory committed:
feat(cache) Add a django cache adapter that reconnects

We often see load-imbalances in our memcache clusters. Application
containers connect through a pool of twemproxy instances, and because
application containers are long-lived and only connect on startup, we
can end up with imbalanced load.

By periodically disconnecting and reconnecting to twemproxy we have
a better chance of getting more even load distribution.

Refs PRODENG-702

constantinius commented on src/sentry/cache/backends/reconnectingmemcache.py:
```diff
@@ -0,0 +1,52 @@
+        if age >= self._reconnect_age:
+            # Close the underlying cache connection if we haven't done that recently.
+            metrics.incr("cache.memcache.reconnect")
+            self._backend.close()
```
I would flip the `if`s and set set `_backend` to `None` when disconnecting.

```suggestion
        if age >= self._reconnect_age:
            # Close the underlying cache connection if we haven't done that recently.
            metrics.incr("cache.memcache.reconnect")
            self._backend.close()
            self._backend = None

        if not self._backend:
            self._backend = PyMemcacheCache(self._server, self._params)
            self._last_reconnect_at = time.time()
```

markstory committed:
Fix mistakes

mwarkentin reviewed (approved):
LGTM, not sure about the cursor comments.

If those aren't a concern lets merge and proceed with some testing.

wedamija reviewed (commented):
I'm wondering if it would make more sense to subclass `PyMemcacheCache`, and then implement the disconnect by overriding `BaseMemcachedCache._cache` instead? Then you don't need to worry about reimplementing every function

markstory committed:
Use class extension instead of decoration

Using a subclass means we can override an internal factory method and be less concered with the public interface of cache adapters.

markstory committed:
Protect against thread races

There is a possibility of multiple threads attempting to reconnect to
memcache concurrently. Use a lock to avoid races.

wedamija commented on src/sentry/cache/backends/reconnectingmemcache.py:
```diff
@@ -0,0 +1,52 @@
+                self._backend.close()
+                self._backend = None
+
+            metrics.incr("cache.memcache.reconnect")
```
Should this stat only fire inside of  `if self._backend`? Otherwise it's counting the first connection as well

markstory committed:
Fix potential deadlock on error
